‘reference to packed field is unaligned’ in nested field of `match` pattern

Consider (https://godbolt.org/z/x39GzMKz4):

```rust
#[repr(packed)]
#[derive(PartialEq)]
pub struct Packed(i32);

enum Y {
    Foo { id: Packed, },
    Bar(i32)
}

fn pred(_: &i32) -> bool {
    return true;
}

fn f(x: Y) {
    match &x {
        Y::Foo { id: Packed(4), .. } => {},
        Y::Bar(s) if pred(s) => {},
        _ => {}
    }
}
```

This results in an error:
```console
error[E0793]: reference to packed field is unaligned
  --> <source>:14:11
   |
14 |     match &x {
   |           ^^
   |
```

I don’t really know what the problem is here, but it might be the match on the `id` field in the first match arm that triggers the error. Moving `Packed(4)` into a local variable works, but if I declare it as a `const` variable instead the error resurfaces. Alternatively, removing the `Y::Bar` match arm for some reason also makes the error go away.

This currently errors on nightly and some testing on godbolt shows that this code results in an error since Rust 1.62. Before that, we still get the same diagnostic; it’s just treated as a warning instead.

--- a/compiler/rustc_middle/src/mir/visit.rs
+++ b/compiler/rustc_middle/src/mir/visit.rs
@@ -1364,13 +1364,13 @@ impl PlaceContext {
         matches!(self, PlaceContext::MutatingUse(MutatingUseContext::Drop))
     }
 
-    /// Returns `true` if this place context represents a borrow.
+    /// Returns `true` if this place context represents a borrow, excluding fake borrows
+    /// (which are an artifact of borrowck and not actually borrows in runtime MIR).
     pub fn is_borrow(self) -> bool {
         matches!(
             self,
-            PlaceContext::NonMutatingUse(
-                NonMutatingUseContext::SharedBorrow | NonMutatingUseContext::FakeBorrow
-            ) | PlaceContext::MutatingUse(MutatingUseContext::Borrow)
+            PlaceContext::NonMutatingUse(NonMutatingUseContext::SharedBorrow)
+                | PlaceContext::MutatingUse(MutatingUseContext::Borrow)
         )
     }
 

--- a/tests/ui/lint/unaligned_references_fake_borrow.rs
+++ b/tests/ui/lint/unaligned_references_fake_borrow.rs
@@ -1,0 +1,27 @@
+//@ check-pass
+
+// Regression test for <https://github.com/rust-lang/rust/issues/137250>.
+
+// Ensure that we don't emit unaligned packed field reference errors for the fake
+// borrows that we generate during match lowering. These fake borrows are there to
+// ensure in *borrow-checking* that we don't modify the value being matched, but
+// they are removed after the MIR is processed by `CleanupPostBorrowck`.
+
+#[repr(packed)]
+pub struct Packed(i32);
+
+fn f(x: Packed) {
+    match &x {
+        Packed(4) => {},
+        _ if true => {},
+        _ => {}
+    }
+
+    match x {
+        Packed(4) => {},
+        _ if true => {},
+        _ => {}
+    }
+}
+
+fn main() {}

